File: web/src/lib/wasm/strava/eta.ts
/**
 * ETA Estimator - Estimates import completion time based on rate limits.
 *
 * Ported from internal/importer/eta.go
 */

export interface ETAState {
  fifteenMinLimit: number
  fifteenMinUsed: number
  fifteenMinReset: number | null // timestamp
  dailyLimit: number
  dailyUsed: number
  dailyReset: number | null // timestamp
  avgCallDuration: number // milliseconds
}

/**
 * Create a new ETA estimator with default values.
 */
export function createETAEstimator(): ETAState {
  return {
    fifteenMinLimit: 100,
    fifteenMinUsed: 0,
    fifteenMinReset: null,
    dailyLimit: 1000,
    dailyUsed: 0,
    dailyReset: null,
    avgCallDuration: 500, // 500ms average
  }
}

/**
 * Update the estimator from rate limit info.
 */
export function updateFromRateLimits(
  state: ETAState,
  usage15Min: number,
  limit15Min: number,
  usageDaily: number,
  limitDaily: number
): void {
  state.fifteenMinUsed = usage15Min
  state.fifteenMinLimit = limit15Min
  state.dailyUsed = usageDaily
  state.dailyLimit = limitDaily
}

/**
 * Get remaining calls in current 15-min window.
 */
function fifteenMinRemaining(state: ETAState): number {
  return state.fifteenMinLimit - state.fifteenMinUsed
}

/**
 * Get remaining calls in current day.
 */
function dailyRemaining(state: ETAState): number {
  return state.dailyLimit - state.dailyUsed
}

/**
 * Estimate time to complete given number of API calls.
 * Returns duration in milliseconds.
 */
export function estimateCompletion(state: ETAState, remainingCalls: number): number {
  if (remainingCalls <= 0) {
    return 0
  }

  const now = Date.now()
  const fifteenMin = 15 * 60 * 1000
  const fifteenMinRemain = fifteenMinRemaining(state)
  const dailyRemain = dailyRemaining(state)

  // If we can complete in current 15-min window
  if (remainingCalls <= fifteenMinRemain && remainingCalls <= dailyRemain) {
    return remainingCalls * state.avgCallDuration
  }

  // Calculate time accounting for rate limit windows
  let totalDuration = 0
  let callsLeft = remainingCalls
  let dailyRem = dailyRemain

  // Use remaining calls in current 15-min window
  if (fifteenMinRemain > 0 && dailyRemain > 0) {
    const usable = Math.min(fifteenMinRemain, dailyRemain, callsLeft)
    totalDuration += usable * state.avgCallDuration
    callsLeft -= usable
    dailyRem -= usable
  }

  if (callsLeft <= 0) {
    return totalDuration
  }

  // Wait for current 15-min window to reset
  if (state.fifteenMinReset && state.fifteenMinReset > now) {
    totalDuration += state.fifteenMinReset - now
  } else {
    totalDuration += fifteenMin
  }

  // Calculate full 15-min windows needed
  const callsPerWindow = Math.min(state.fifteenMinLimit, dailyRem)
  if (callsPerWindow <= 0) {
    // Daily limit exhausted, need to wait for daily reset
    if (state.dailyReset && state.dailyReset > now) {
      return totalDuration + (state.dailyReset - now)
    }
    return totalDuration + 24 * 60 * 60 * 1000
  }

  const fullWindows = Math.floor(callsLeft / callsPerWindow)
  totalDuration += fullWindows * fifteenMin
  callsLeft -= fullWindows * callsPerWindow

  // Partial final window
  if (callsLeft > 0) {
    totalDuration += callsLeft * state.avgCallDuration
  }

  return totalDuration
}

/**
 * Format the ETA duration as human-readable string.
 */
export function formatETA(durationMs: number): string {
  if (durationMs <= 0) {
    return '< 1 min'
  }

  const hours = Math.floor(durationMs / (60 * 60 * 1000))
  const minutes = Math.floor((durationMs % (60 * 60 * 1000)) / (60 * 1000))

  if (hours > 24) {
    const days = Math.floor(hours / 24)
    const remainingHours = hours % 24
    if (remainingHours > 0) {
      return `${formatPlural(days, 'day')} ${formatPlural(remainingHours, 'hour')}`
    }
    return formatPlural(days, 'day')
  }

  if (hours > 0) {
    if (minutes > 0) {
      return `${formatPlural(hours, 'hour')} ${formatPlural(minutes, 'min')}`
    }
    return formatPlural(hours, 'hour')
  }

  if (minutes > 0) {
    return formatPlural(minutes, 'min')
  }

  return '< 1 min'
}

function formatPlural(n: number, unit: string): string {
  if (n === 1) {
    return `1 ${unit}`
  }
  return `${n} ${unit}s`
}

File: internal/importer/eta.go
// Package importer handles importing data from Strava.
package importer

import (
	"fmt"
	"time"
)

// ETAEstimator estimates completion time based on rate limits.
type ETAEstimator struct {
	// Rate limits (from Strava)
	FifteenMinLimit int
	FifteenMinUsed  int
	FifteenMinReset time.Time
	DailyLimit      int
	DailyUsed       int
	DailyReset      time.Time

	// Average API call duration
	AvgCallDuration time.Duration
}

// NewETAEstimator creates a new ETA estimator with default values.
func NewETAEstimator() *ETAEstimator {
	return &ETAEstimator{
		FifteenMinLimit: 100,
		DailyLimit:      1000,
		AvgCallDuration: 500 * time.Millisecond,
	}
}

// UpdateFromRateLimits updates the estimator from rate limit info.
func (e *ETAEstimator) UpdateFromRateLimits(fifteenMinUsed, fifteenMinLimit int, fifteenMinReset time.Time, dailyUsed, dailyLimit int, dailyReset time.Time) {
	e.FifteenMinUsed = fifteenMinUsed
	e.FifteenMinLimit = fifteenMinLimit
	e.FifteenMinReset = fifteenMinReset
	e.DailyUsed = dailyUsed
	e.DailyLimit = dailyLimit
	e.DailyReset = dailyReset
}

// FifteenMinRemaining returns remaining calls in current 15-min window.
func (e *ETAEstimator) FifteenMinRemaining() int {
	return e.FifteenMinLimit - e.FifteenMinUsed
}

// DailyRemaining returns remaining calls in current day.
func (e *ETAEstimator) DailyRemaining() int {
	return e.DailyLimit - e.DailyUsed
}

// EstimateCompletion estimates time to complete given number of API calls.
func (e *ETAEstimator) EstimateCompletion(remainingCalls int) time.Duration {
	if remainingCalls <= 0 {
		return 0
	}

	now := time.Now()
	fifteenMinRemain := e.FifteenMinRemaining()
	dailyRemain := e.DailyRemaining()

	// If we can complete in current 15-min window
	if remainingCalls <= fifteenMinRemain && remainingCalls <= dailyRemain {
		return time.Duration(remainingCalls) * e.AvgCallDuration
	}

	// Calculate time accounting for rate limit windows
	totalDuration := time.Duration(0)
	callsLeft := remainingCalls

	// Use remaining calls in current 15-min window
	if fifteenMinRemain > 0 && dailyRemain > 0 {
		usable := min(fifteenMinRemain, dailyRemain, callsLeft)
		totalDuration += time.Duration(usable) * e.AvgCallDuration
		callsLeft -= usable
		dailyRemain -= usable
	}

	if callsLeft <= 0 {
		return totalDuration
	}

	// Wait for current 15-min window to reset
	if !e.FifteenMinReset.IsZero() && e.FifteenMinReset.After(now) {
		totalDuration += e.FifteenMinReset.Sub(now)
	} else {
		totalDuration += 15 * time.Minute
	}

	// Calculate full 15-min windows needed
	callsPerWindow := min(e.FifteenMinLimit, dailyRemain)
	if callsPerWindow <= 0 {
		// Daily limit exhausted, need to wait for daily reset
		if !e.DailyReset.IsZero() && e.DailyReset.After(now) {
			return totalDuration + e.DailyReset.Sub(now)
		}
		return totalDuration + 24*time.Hour
	}

	fullWindows := callsLeft / callsPerWindow
	totalDuration += time.Duration(fullWindows) * 15 * time.Minute
	callsLeft -= fullWindows * callsPerWindow

	// Partial final window
	if callsLeft > 0 {
		totalDuration += time.Duration(callsLeft) * e.AvgCallDuration
	}

	return totalDuration
}

// FormatETA formats the ETA duration as human-readable string.
func FormatETA(d time.Duration) string {
	if d <= 0 {
		return "< 1 min"
	}

	hours := int(d.Hours())
	minutes := int(d.Minutes()) % 60

	if hours > 24 {
		days := hours / 24
		hours %= 24
		if hours > 0 {
			return formatPlural(days, "day") + " " + formatPlural(hours, "hour")
		}
		return formatPlural(days, "day")
	}

	if hours > 0 {
		if minutes > 0 {
			return formatPlural(hours, "hour") + " " + formatPlural(minutes, "min")
		}
		return formatPlural(hours, "hour")
	}

	if minutes > 0 {
		return formatPlural(minutes, "min")
	}

	return "< 1 min"
}

func formatPlural(n int, unit string) string {
	if n == 1 {
		return "1 " + unit
	}
	return fmt.Sprintf("%d %ss", n, unit)
}
